mixnet/client.py

```diff
@@ -0,0 +1,75 @@
+from __future__ import annotations
+
+import queue
+import time
+from datetime import datetime, timedelta
+from threading import Thread
+
+from mixnet.mixnet import Mixnet, MixnetTopology
+from mixnet.node import PacketQueue
+from mixnet.packet import PacketBuilder
+from mixnet.poisson import poisson_interval_sec
+
+
+class MixClientRunner(Thread):
+    """
+    Emit packets at the Poisson emission_rate_per_min.
+
+    If a real packet is scheduled to be sent, this thread sends the real packet to the mixnet,
+    and schedules redundant real packets to be emitted in the next turns.
+
+    If no real packet is not scheduled, this thread emits a cover packet according to the emission_rate_per_min.
+    """
+
+    def __init__(
+        self,
+        mixnet: Mixnet,
+        topology: MixnetTopology,
+        emission_rate_per_min: int,  # Poisson rate parameter: lambda in the spec
+        redundancy: int,  # b in the spec
+        real_packet_queue: PacketQueue,
+        outbound_socket: PacketQueue,
+    ):
+        super().__init__()
+        self.mixnet = mixnet
+        self.topology = topology
+        self.emission_rate_per_min = emission_rate_per_min
+        self.redundancy = redundancy
+        self.real_packet_queue = real_packet_queue
+        self.redundant_real_packet_queue: PacketQueue = queue.Queue()
+        self.outbound_socket = outbound_socket
+
+    def run(self) -> None:
+        # Here in Python, this thread is implemented in synchronous manner.
+        # In the real implementation, consider implementing this in asynchronous if possible.
```

Comment:
Yes, here I see clearly that this is meant to be a long running task. Is this where you [were referring ](https://github.com/logos-co/nomos-specs/pull/49#discussion_r1457117373) about changing to async?

Reply:
Yes, exactly. This thread has many short sleeps to minimize the time distortion(?), compared to having a long sleep, such as `time.sleep(poisson_interval_sec(self.emission_rate_per_min)`, because there might be some computation time to process packets. I actually don't like this approach. Also, I guess this can be more readable if we use async timer.